syft/pkg/cataloger/internal/dependency/resolver.go

```diff
@@ -0,0 +1,81 @@
+package dependency
+
+import (
+	"sort"
+
+	"github.com/scylladb/go-set/strset"
+
+	"github.com/anchore/syft/syft/artifact"
+	"github.com/anchore/syft/syft/pkg"
+	"github.com/anchore/syft/syft/pkg/cataloger/generic"
+)
+
+// Prosumer is a producer and consumer, in this context, for packages that provide resources and require resources.
+type Prosumer interface {
```

Comment:
After reflecting on this for a bit, I would like to suggest removing the `Prosumer` interface in favor of 2 callback functions: one for getting IDs produced by a package and another getting IDs depended on. I believe that would reduce the complexity and make things more composable. The biggest draw for using the `generic.Cataloger` is that a user only has to implement a simple function, there aren't interfaces to implement. I feel that adding something requiring implementing interfaces here that is likely to be used by quite a few `generic.Cataloger` users is a bit of a step backwards in terms of usability. And the user could simply implement a `pkg.Cataloger` and there would not really be as much a need for this because they would have a list of all the packages to be able to construct dependencies on fairly simply. By keeping with the function-only usage, I think it would be more consistent with the intended users of this functionality.

I would envision it looking something like:
```
type Producer func(pkg.Package) []string

type RelationshipResolver struct {
	provides Producer
	depends Producer
}

func NewRelationshipResolver(ids Producer, dependencies Producer) generic.Processor { ...
```

Usage, is still simple, e.g.:
```
WithProcessors(dependency.Processor(idProvider, dependencyProvider)) // where idProvider and dependencyProvider are local functions to provide ids (already implemented as part of a prosumer in this PR)
```
This would make certain types of dependency mappings only need to implement one function, and potentially look like:
```
WithProcessors(dependency.Processor(dependency.NameVersionProvider, dependencyProvider))
```

And would eliminate the duplicated boilerplate code such as:
```
var _ dependency.Prosumer = (*dpkgDBEntryProsumer)(nil)

type dpkgDBEntryProsumer struct{}

func newDBProsumer() dpkgDBEntryProsumer {
	return dpkgDBEntryProsumer{}
}
```

Reply:
Actually, I started with something that looked exactly like this (different name though), but when working with it something about it bothered me:
```
WithProcessors(dependency.Processor(idProvider, dependencyProvider))
```
vs
```
WithProcessors(dependency.Processor(dependencyProvider, idProvider))
```
would be allowed from a type point of view, but the latter is a mistake and will result in incorrect behavior. Ideally this would be as clear as possible about what is being passed and also difficult to misuse.

I then went in this direction:
```
type Producer func(pkg.Package) []string
type Consumer func(pkg.Package) []string


type RelationshipResolver struct {
	provides Producer
	depends Consumer
}
```

But then realized that this doesn't help in most cases since we're passing function literals... so the same problem would arise!

After more thought it seemed natural to lean into either one or two interfaces so what is being passed is clear from a type perspective as early as compile time, with the added benefit that the function names semantically hint at the intended behavior.

Reply:
If someone is using this feature, they are going to need to know what each of the required functions do whether they're interfaces or separate functions. I won't disagree that having an interface that enforces 2 separate functions to be implemented is safer, to some degree. But I definitely think it's less convenient, especially for the case where packages could be identified using a common `name@version` string or something similar where the only requirement is to provide a function for _dependencies_, not _provides_. This could be wrapped up in yet another implementation in the `dependencies` package but that's more complexity because different catalogers are doing different things. If you feel strongly this is the right way to go, then that's ok with me 👍

Reply:
from an offline conversation we've found a good alternative approach that balances many of the comments. Short answer is replace multiple function calls / an interface with a single `func(pkg.Package) dependency.Specification` (where the specification has both provides and depends data). See the code for the full final state.